DOCSP-13840 crud search text (#76)
mongodb/docs-golang: source/fundamentals/crud/read-operations/text.txt

``movies`` collection of the ``marvel`` database with the following
snippet:

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/textSearch.go
   :language: go
   :dedent:
   :start-after: begin insert docs
   :end-before: end insert docs

.. include:: /includes/fundamentals/automatic-db-coll-creation.rst

Each document contains the name and release year of the Marvel movie
that corresponds to the ``title`` and ``year`` fields.

.. include:: /includes/fundamentals/truncated-id.rst

Text Index
~~~~~~~~~~

You must create a **text index** before running a text search. A text
index specifies the string or string array field to run a text
search on.

The examples in the following sections run text searches on the
``title`` field in the ``movies`` collection. To enable text searches on
the ``title`` field, create a text index using with the following snippet: 

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/textSearch.go
   :language: go
   :dedent:
   :start-after: begin text index
   :end-before: end text index

.. _text-search-go:

Text Search
-----------

A text search retrieves documents that contain a **term** or a

> **phrase** in the text indexed fields. A term is a sequence of

Author: I think removing "text" makes the sentence incorrect because a user can have non-text indexes too.

characters that excludes whitespace characters. A phrase is a sequence
of terms with any number of whitespace characters.

To perform a text search, use the ``$text`` evaluation query operator,
followed by the ``$search`` field in your query filter. The ``$text`` operator
performs a text search on the text indexed fields. The ``$search`` field
specifies the text to search in the text indexed fields.

Query filters for text searches use the following format:

.. code-block:: go

   filter := bson.D{{"$text", bson.D{{"$search", "<text to search>"}}}}

.. _term-search-go:

Search by a Term
~~~~~~~~~~~~~~~~

To search for a term, specify the term as a string in your query filter.
To search for multiple terms, separate each term with spaces in the string.

.. note::

   When searching for multiple terms, the ``Find()`` function returns
   documents with at least one of the terms in text indexed fields.

Example
```````

The following example runs a text search for titles that contain the term "War":

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/textSearch.go
   :language: go
   :dedent:
   :start-after: begin term search
   :end-before: end term search

> After running this example, the output resembles the following:

Author: I prefer keeping "resembles" because it's not an exact match -- the ObjectID is truncated and will vary for all users.

.. code-block:: none
   :copyable: false

   [{_id ObjectID("...")} {title Avengers: Infinity War} {year 2018}]
   [{_id ObjectID("...")} {title Captain America: Civil War} {year 2016}]

Search by a Phrase
~~~~~~~~~~~~~~~~~~

To search for a phrase, specify the phrase with escaped quotes as a
string in your query filter. If you don't add escaped quotes around the

> phrase, the ``Find()`` function runs a :ref:`term search <term-search-go>`.

Reviewer: Ah... so, if term search allows multiple terms separated by spaces within a single string to search for individual terms, please ignore my above comment.

.. tip::

   Escaped quotes are a backslash character followed by a double quote
   character.

Example
```````

The following example runs a text search for titles that contain the
phrase "Infinity War":

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/textSearch.go
   :language: go
   :dedent:
   :start-after: begin phrase search
   :end-before: end phrase search

> After running this example, the output resembles the following:

Reviewer: Same comment here.

.. code-block:: none
   :copyable: false

   [{_id ObjectID("...")} {title Avengers: Infinity War} {year 2018}]

Search with Terms Excluded
~~~~~~~~~~~~~~~~~~~~~~~~~~

> For each term you want to exclude from your text search, specify the

Reviewer: ```suggestion
For each term you want to exclude from your text search, 
```

> term prefixed with a minus sign as a string in your query filter.

Reviewer: ```suggestion
in your query filter, prefix the term or phrase in the string with a minus sign (``-``).
```

.. important::

   You must search for at least one term if you want to exclude
   terms from your search. If you don't search for any terms, the

>    ``Find()`` function doesn't return any documents.

Reviewer: I suggest replacing the note above with something similar to the following: 

Don't exclude all terms in your text search. You must include at least one term to search so that the ``Find()`` function can return documents.

Reviewer: I disagree. To me, it reads as though the user must do a search for terms to exclude terms, which is not the same as all terms in the search can't be excluded only terms (OR) one or more terms should be non-excluded terms (or terms without the minus prefix).

Example
```````

The following example runs a text search for titles that contain the
phrase "Captain America", but does not contain the term "Avenger":

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/textSearch.go
   :language: go
   :dedent:
   :start-after: begin exclude term search
   :end-before: end exclude term search

> After running this example, the output resembles the following:

Reviewer: Same comment here.

.. code-block:: none
   :copyable: false

   [{_id ObjectID("...")} {title Captain America: The Winter Soldier} {year 2014}]
   [{_id ObjectID("...")} {title Captain America: Civil War} {year 2016}]

Sort by Relevance
~~~~~~~~~~~~~~~~~

A text search assigns a numerical text score to indicate how closely
each result matches the string in your query filter. To reveal the text
score in your output, use a projection to retrieve the "textScore"
metadata. You can sort the text score in descending order by specifying
a sort on the "textScore" metadata.

Example
```````

The following example performs the following actions:

- Runs a text search for titles that contain the term "Avenger"
- Sorts the results in descending order based on their text score
- Includes the ``title`` and ``score`` fields from the results

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/textSearch.go
   :language: go
   :dedent:
   :start-after: begin text score
   :end-before: end text score

> After running this example, the output resembles the following:

Reviewer: Same comment here.

.. code-block:: none
   :copyable: false

   [{title The Avengers} {score 1}]
   [{title Avengers: Infinity War} {score 0.6666666666666666}]
   [{title Captain America: The First Avenger} {score 0.625}]

Aggregation
-----------

You can also include the ``$text`` evaluation query operator in the
:manual:`$match </reference/operator/aggregation/match/>` stage to
perform a text search in an aggregation pipeline.

Term Search Example
~~~~~~~~~~~~~~~~~~~

The following example runs a text search for titles that contain the term "Winter":

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/textSearch.go
   :language: go
   :dedent:
   :start-after: begin aggregate text search
   :end-before: end aggregate text search

> After running this example, the output resembles the following:

Reviewer: Same comment here.

.. code-block:: none
   :copyable: false

   [{_id ObjectID("...")} {title Captain America: The Winter Soldier} {year 2014}]

Text Score Example
~~~~~~~~~~~~~~~~~~

The following example performs the following actions:

- Runs a text search for titles that contain the term "Avenger"
- Sorts the results in descending order based on their text score
- Includes the ``title`` and ``score`` fields from the results

.. literalinclude:: /includes/fundamentals/code-snippets/CRUD/textSearch.go
   :language: go
   :dedent:
   :start-after: begin aggregate text score
   :end-before: end aggregate text score

> After running this example, the output resembles the following:

Reviewer: Same comment here.

.. code-block:: none
   :copyable: false

   [{title The Avengers} {score 1}]
   [{title Avengers: Infinity War} {score 0.6666666666666666}]
   [{title Captain America: The First Avenger} {score 0.625}]

Additional Information
----------------------

> For more information on the operations mentioned, see the following

Reviewer: ```suggestion
For more information on the operations used in the examples in this guide, see the following
```

guides:

> - :doc:`Specify a Query </fundamentals/crud/query-document>`

Reviewer: Should this be:

```suggestion
- :doc:`filter </fundamentals/crud/query-document>`
```

Reviewer: Suggest rephrasing the link text to be consistent with the last two bullet items

> - :doc:`Sort Results </fundamentals/crud/read-operations/sort>`

Reviewer: Should this be: 

```suggestion
- :doc:`$sort </fundamentals/crud/read-operations/sort>`
```

> - :doc:`Specify Which Fields to Return </fundamentals/crud/read-operations/project>`

Reviewer: Should this be: 

```suggestion
- :doc:`$project </fundamentals/crud/read-operations/project>`
```

===========
Search Text
===========

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

> In this guide, you can learn how to run a :ref:`text search <text-search-go>`.

Reviewer 2: Optional: Consider adding a clarifying note that MongoDB text search is different than [Atlas Search](https://docs.atlas.mongodb.com/atlas-search/), which has significantly different behavior and expanded functionality.
